### PR TITLE
medusa plugin

### PR DESCRIPTION
**Pull request by Marquez96**
### Motivation for changes:
I was updating myinfrastructure with medusa, and i tried to connect with flexget via sickbear plugin, but the api of sickbear of medusa was outdated and i create a new plugin specific for medusa

### Detailed changes:
I created the plugin based on old sickbear plugin, i reformat and created all the rest requests of that

### Implemented feature requests:
- Feathub https://feathub.com/Flexget/Flexget/+65

### Config usage if relevant (new plugin or updated schema):
```

templates:
  medusa:
    configure_series:
      from:
        medusa:
          base_url: http://127.0.0.1
          port: 8081
          username: username
          password: password
          only_monitored: True

# ---------- General config
    configure_series:
      from:
        medusa:
          base_url: string
          port: string default 8081
          username: string
          password: string
          only_monitored: boolean default False
          included_ended: boolean default False
```
### Log and/or tests output (preferably both):
```
DEBUG MODE 
2018-12-21 23:24 DEBUG    medusa        updating_series processing show: {'title': 'Westworld', 'airs': 'Sunday 0:00AM', 'showType': 'series', 'rating': {'imdb': {'votes': 317083, 'rating': '8.8'}}, 'nextAirDate': None, 'cache': {'banner': '/opt/medusa/cache/images/tmdb/63247.poster.jpg', 'poster': '/opt/medusa/cache/images/tmdb/63247.poster.jpg'}, 'language': 'es', 'genres': ['Drama', 'Mystery', 'Western', 'Ciencia ficción', 'Science-Fiction'], 'id': {'slug': 'tmdb63247', 'tmdb': 63247, 'imdb': 'tt0475784', 'trakt': None}, 'classification': '', 'config': {'dvdOrder': False, 'sports': False, 'airByDate': False, 'qualities': {'allowed': [2, 4, 8, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768], 'preferred': []}, 'paused': False, 'defaultEpisodeStatus': 'Wanted', 'anime': False, 'release': {'requiredWords': [], 'ignoredWords': []}, 'aliases': [], 'subtitlesEnabled': False, 'location': '/mnt/media/series/westworld', 'seasonFolders': False, 'scene': False}, 'network': 'HBO', 'year': {'start': 2016}, 'countries': ['UNITED STATES'], 'type': 'Scripted', 'runtime': 62, 'akas': {}, 'plot': 'Westworld ...', 'country_codes': ['us'], 'status': 'Continuing', 'indexer': 'tmdb'}

INFO MODE - Silence
2018-12-20 20:04 VERBOSE  details       updating_series Produced 16 entries.


```
#### To Do:
It's recomendable to detect qualities in medusa, now i only can choose for only_monitories and ended series

### CODE REVIEW

File: flexget/plugins/input/medusa.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+            port=config.get('port'),
+        )
+
+        body_auth = dict(
+            username=config.get('username'),
+            password=config.get('password'),
+        )
+
+        try:
+            api_key = json.loads(task.requests.post(
```
Comment (liiight) [inline flexget/plugins/input/medusa.py]: you can use native `requests` formatting here:
```suggestion
        api_key = task.requests.post('{}/authenticate'.format(base_url), json=body_auth, headers=self.headers).json()
```

File: flexget/plugins/input/medusa.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                '{}/authenticate'.format(base_url),
+                data=json.dumps(body_auth),
+                headers=self.headers
+            ).content.decode('utf-8'))['token']
+
+            self.headers.update(
+                authorization='Bearer ' + api_key
+            )
+
+            series = json.loads(task.requests.get(
```
Comment (liiight) [inline flexget/plugins/input/medusa.py]: look above for optimal request syntax. Also, ideally pass url param via its own dict:
```suggestion
			params = {
				'limit': 1000
			}
            series = task.requests.get('{}/series'.format(base_url), params=params).json()
```

File: flexget/plugins/input/medusa.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+from flexget.entry import Entry
+
+log = logging.getLogger('medusa')
+
+
+class Medusa(object):
+    schema = {
+        'type': 'object',
+        'properties': {
+            'base_url': {'type': 'string'},
```
Comment (liiight) [inline flexget/plugins/input/medusa.py]: you can set expected format to `uri`:
```suggestion
            'base_url': {'type': 'string', 'format': 'uri'},
```

File: flexget/plugins/input/medusa.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            ).content.decode('utf-8'))
+        except RequestException as e:
+            raise plugin.PluginError(
+                'Unable to get series in Medusa at {}://{}:{}{}. Error: {}'.format(
+                    parsed_url.scheme, parsed_url.netloc, config.get('port'), parsed_url.path, e)
+            )
+
+        entries = []
+        for show in series:
+            log.debug('processing show: {}'.format(show))
```
Comment (liiight) [inline flexget/plugins/input/medusa.py]: let logger do its own string interpolation:
```suggestion
            log.debug('processing show: %s', show)
```

**Comment by Marquez96** [comment]
The requested changes have been made, 

Thank you for those observations

File: flexget/plugins/input/medusa.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            'username': {'type': 'string'},
+            'password': {'type': 'string'},
+            'only_monitored': {'type': 'boolean', 'default': False},
+            'include_ended': {'type': 'boolean', 'default': False},
+        },
+        'required': ['username', 'password', 'base_url'],
+        'additionalProperties': False
+    }
+
+    headers = {
```
Comment (liiight) [inline flexget/plugins/input/medusa.py]: this is redundant, as it is implicitly added when you use the `json` keyword when sending requests

File: flexget/plugins/input/medusa.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+            params=params,
+            headers=self.headers
+        ).json()
+
+        entries = []
+        for show in series:
+            log.debug('processing show: %s', show)
+            if (show['config']['paused'] and config.get('only_monitored')) or \
+                    show['status'] == 'Ended' and not config.get('include_ended'):
+                continue
```
Comment (liiight) [inline flexget/plugins/input/medusa.py]: maybe add a debug message here?

**Comment by Marquez96** [comment]
Debug message added and remove the default headers

**Comment by liiight** [comment]
Thanks! Can you please update the wiki?

**Comment by Marquez96** [comment]
Yes, of course,
https://flexget.com/Plugins/Medusa

**Pull request merged**